client/rpc-servers/src/lib.rs

```diff
@@ -70,9 +70,9 @@ pub struct Config<'a, M: Send + Sync + 'static> {
 }
 
 /// Start RPC server listening on given address.
-pub async fn start_server<M: Send + Sync + 'static>(
+pub async fn start_server_and_get_socket_address<M: Send + Sync + 'static>(
 	config: Config<'_, M>,
-) -> Result<ServerHandle, Box<dyn StdError + Send + Sync>> {
+) -> Result<(ServerHandle, SocketAddr), Box<dyn StdError + Send + Sync>> {
```

Comment:
nit: Not a fan of return tuples, we might want to expand and use https://github.com/gensyn-ai/substrate/pull/5/files#diff-3f0c43f8bc417719ed28ca9d9d2cb367e4330242a5df4d289ba89494be4cfc10R45 as a type holding both the `ServerHandle` and the `SocketAddr`

Reply:
Like

```rust
struct Server {
    handle: ServerHandle,
    listen_addr: SocketAddr,
}
```

Might work, but might also clash, as type aliases in Rust are just syntactic sugar, not actual types.

Reply:
I'd also suggest to keep the method name in this case. Furthermore, we could consider keeping the old "automatic" behavior and return `Option<SocketAddr>` (right now we turned that into an error).

That'd keep source-level API compatibility, for the most part.

Reply:
> Might work, but might also clash, as type aliases in Rust are just syntactic sugar, not actual types.

We can sidestep all issues having `DerefMut` and `Deref` implemented. I think having `Server` as type alias was a mistake from the get-go (not ours though).

Reply:
Isn't that a little scandalous though? Deref polymorphism via implementing `Deref` for types that are not smart pointers is a [frowned upon affair](https://doc.rust-lang.org/std/ops/trait.Deref.html), I tend to err on the side of not implementing it.

Reply:
One concern I have with this is that ```Server``` is a public type, so changing it to a struct may break other library consumers (outside of Gensyn) if we upstream this PR.

Reply:
That's true. Question is whether that should be a concern or not.

Reply:
Take a look at how I addressed this. I created a new ServerAndListenAddress type for the public crate API, and continued to use a return tuple for the private internal function (so I didn't have to make yet another new type). Note that [here](https://github.com/gensyn-ai/substrate/blob/6af14f3bb4b340fb4a7b4ad4ca47cdff74e5b932/client/service/src/lib.rs?plain=1#L426) the Server type is changing, so I can't reuse my ServerAndListenAddress type.